File: Maps/Subarray Sum Equals K.cpp
/*Given an array of integers nums and an integer k, return the total number of continuous subarrays whose sum equals to k.

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2
Example 2:

Input: nums = [1,2,3], k = 3
Output: 2
*/

class Solution {
public:
    int subarraySum(vector<int>& nums, int k)
    {
       int count=0;
        int sum=0;
        unordered_map<int,int>mp={{0,1}};
      for(int i=0;i<nums.size();i++)
      {
          sum=sum+nums[i];
          
          if(mp.find(sum-k)!=mp.end())
          {
              count=count+mp[sum-k];
          }
          
          mp[sum]++;
      }
        return count;
    }
};
 

File: PrefixSum/Subarray Sum Equals K.java
class Solution {
    public int subarraySum(int[] nums, int k) {
      
        Map<Integer, Integer> mp = new HashMap<>();

        mp.put(0,1);

        int count = 0;
        int cumSum = 0;

        for(int i=0;i<nums.length;i++){
            
            cumSum += nums[i];

            if(mp.containsKey(cumSum-k)){
                 count += mp.get(cumSum-k);
            }

            mp.put(cumSum, mp.getOrDefault(cumSum,0)+1);
        }

        return count;


    }
}
